aptos-move/framework/aptos-framework/sources/vesting.move

```diff
@@ -0,0 +1,451 @@
+    /// epoch to terminate.
+    const EPENDING_ACTIVE_STAKE: u64 = 12;
+    /// Can only withdraw from terminated vesting pools.
+    const EVESTING_POOL_NOT_TERMINATED: u64 = 13;
+    /// There's no unlocked stake to withdraw from the stake pool.
+    const ENO_STAKE_TO_WITHDRAW: u64 = 14;
+    /// Vesting has not started yet. Need to wait until the specified vesting start timestamp.
+    const EVESTING_HAS_NOT_STARTED: u64 = 15;
+    /// No new vesting period has passed since the last distribution.
+    const ENO_NEW_VESTING_PERIOD_HAS_PASSED: u64 = 16;
+    /// There are no vested coins and rewards to distribute.
+    const ENO_COINS_TO_DISTRIBUTE: u64 = 17;
+    /// Shareholder not found in specified vesting pool.
+    const ESHAREHOLDER_NOT_FOUND: u64 = 18;
+
+    struct VestingPool has key {
+        state: u64,
+        admin: address,
+        beneficiaries: SimpleMap<address, address>,
+        shareholders: vector<address>,
+        shares: SimpleMap<address, u64>,
+        // Total amount at the start of the vesting pool.
+        grant_amount: u64,
+        // Withdrawal address where all funds would be released back to if the admin terminates the vesting contract.
+        // This is for security purposes in case the admin account is different from where funds should be held.
+        withdrawal_address: address,
+        // The owner cap for the stake pool where the unvested funds will be staked at.
+        stake_pool_owner_cap: OwnerCapability,
+        // Where the vesting pool is located at. Included for convenience.
+        pool_address: address,
+
+        // The vesting schedule as a list of fractions that vest for each period. The last number is repeated until the
+        // vesting amount runs out.
+        // For example [1/24, 1/24, 1/48] with a period of 1 month means that after vesting starts, the first two months
+        // will vest 1/24 of the original total amount. From the third month only, 1/48 will vest until the vesting fund
+        // runs out.
+        vesting_schedule: vector<FixedPoint32>,
+        // When the vesting should start.
+        vesting_start_timestamp_secs: u64,
+        // How long each vesting period is. For example 1 month.
+        vesting_period_duration: u64,
+        // Last vesting period, 1-indexed. For example if 2 months have passed, the last vesting period, if distribution
+        // was requested, would be 2. Default value is 0 which means there have been no vesting periods yet.
+        last_vesting_period: u64,
+
+        // Not used currently but might in the future if this module is upgraded. We want to keep the signer cap just
+        // in case.
+        signer_cap: SignerCapability,
+    }
+
+    /// Resource holding a nonce that's stored at the admin account. We'll use this to create resource accounts for new
+    /// vesting pools so there's no address collision.
+    struct AdminNonce has key {
+        vesting_pool_creation_nonce: u64,
+    }
```

Comment:
I don't think that this is necessary, we can just use guid if you really need it. Again someone can pre-compute these and create hell for you.

Reply:
This is just for convenience as:
1. In genesis or in any scenario where multiple vesting contracts are created in a single tx, account sequence number won't be incremented after each vesting contract creation.
2. Address squatting is a similar challenge to in staking_contract, so create_vesting_contract accepts an optional salt that the caller can pass in to get around address squatting.